### PR TITLE
feat(replays): Create the new DOM tab for Replays

### PR DESCRIPTION
**Pull request by ryan953**
Adds a new DOM table where we can see a table listing each breadcrumb that has interacted with the DOM. For each crumb we can see the actual HTML snippet that the breadcrumb refers to, so for a 'click' event we can see the full html of the button that was clicked.

Screen Shot 2022-07-20 at 3 16 17 PM

This PR has minimal styling for the table so far. It can improve with help from:
- #36587 should give an example for base table styles, so all our replay tables look part of the family
- #36880 promises to supply an html formatter for use

The react hook seems to be quick enough on the examples i've tested.  RRweb has an option [`useVirtualDom`](https://github.com/rrweb-io/rrweb/blob/master/guide.md#options-1) which should be kicking in on the line `replayerRef.pause(replay.getEvent().endTimestamp);` because that line we're asking to scrub right to the end of the video, so no need to actual change the live dom.
I tried to construct the Replayer instance without attached `domRoot` to the document, but it wasn't possible, there are some codepaths inside rrweb that call `scrollTo` which are failing. An upstream fix for that is https://github.com/rrweb-io/rrweb/pull/939. If that gets merged then things might get more isolated, but it's not required afaik now.

Relates to: #36892

### CODE REVIEW

**Comment by billyvg** [comment]
👏👏👏 This is super neat -- some design feedback:

* The oval icon here looks awkward and the text is squished: 
image
* The selector text should probably be monospaced


Future things:
* copy to clipboard for selector
* hover effect (like in breadcrumbs)


Other:
* Nav bar tab for `DOM` is a bit hard to click (it's only 30px wide) -- could use some padding

File: static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import {useEffect, useState} from 'react';
+import first from 'lodash/first';
+import {Replayer} from 'rrweb'; // , ReplayerEvents
```
Comment (billyvg) [inline static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx:3]:
```suggestion
import {Replayer} from 'rrweb';
```

File: static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import {eventWithTime} from 'rrweb/typings/types';
+
+import type {Crumb} from 'sentry/types/breadcrumbs';
+import type ReplayReader from 'sentry/utils/replays/replayReader';
+
+const IncrementalSnapshot = 3;
```
Comment (billyvg) [inline static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx:9]: You could use the enum https://github.com/rrweb-io/rrweb/blob/master/packages/rrweb/src/types.ts#L18

File: static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+type Extraction = {
+  crumb: Crumb;
+  html: string;
+  timestamp: number;
+};
+
+type HookOpts = {
+  domRoot: null | HTMLDivElement;
```
Comment (ryan953) [inline static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx:18]: I think i can remove this opt and just let the hook create it's own hidden div on the `<body>` instead.

File: static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx (hunk 2 above)
Comment (ryan953) [reply static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx]: I can't sort out how to import it. both `import {EventType} from 'rrweb/typings';` and `import {EventType} from 'rrweb/typings/types';` fail like this:
Screen Shot 2022-07-21 at 2 06 09 PM
 
the `import {eventWithTime} ...` does work, but that's a TS only import, not webpack.

So I've copied in the whole enum and commented its origin. Updating to rrweb 2.0 might be fun here.

**Comment by ryan953** [comment]
For the design feedback:
> The oval icon here looks awkward and the text is squished:
> The selector text should probably be monospaced
> copy to clipboard for selector
> hover effect (like in breadcrumbs)
> Nav bar tab for DOM is a bit hard to click (it's only 30px wide) -- could use some padding

I want to punt it all. Context is we had an internal show'n tell / preview and some good ideas came out, like make it a list, selector under the icon, etc.
I'd like to do a single bigger iteration taking those ideas into account as well.


_Also, idk why the purple icon isn't a circle. It's annoying._

File: static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx (hunk 3 above)
Comment (ryan953) [reply static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx]: done

File: static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+      speed: 99999,
+      skipInactive: true,
+      triggerFocus: false,
+      plugins: [
+        new BreadcrumbReferencesPlugin({
+          crumbs,
+          isFinished: isLastRRWebEvent,
+          onFinish: rows => {
+            setBreadcrumbReferences(rows);
+            document.body.removeChild(domRoot);
```
Comment (billyvg) [inline static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx:73]: I suppose it's quite unlikely that the useEffect cleanup function gets called before the replay finishes?

**Review by billyvg — APPROVED** [review]

File: static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    const lastTimestamp = lastEvent.timestamp;
+
+    const isLastRRWebEvent = (event: eventWithTime) => lastTimestamp === event.timestamp;
+
+    const replayerRef = new Replayer(rrwebEvents, {
+      root: domRoot,
+      loadTimeout: 1,
+      showWarning: false,
+      blockClass: 'sr-block',
+      speed: 99999,
```
Comment (ryan953) [inline static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx:66]: this is random

File: static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx (hunk 4 above)
Comment (ryan953) [reply static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx:77]: It's all async with timers afaik, so useEffect could get cleaned up before the replay finishes yeah. 

The issue we could see is the `setBreadcrumbReferences` call is made, but the react component is unmounted so the state change can't cause the re-render. That's solvable with a flag i guess.

**Pull request merged**